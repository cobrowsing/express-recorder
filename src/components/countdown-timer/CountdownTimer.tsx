import { Component, h } from "preact";
const styles = require("./style.scss");

type Props = {
    initialValue: number;
    onCountdownComplete: () => void;
    radius?: number;
};

type State = {
    countdown: number;
};

/**
 * Component to show countdown from X to 0. uses for delay between clicking on start recording to the actual recording
 */
export class CountdownTimer extends Component<Props, State> {
    static defaultProps = {
        radius: 44
    };

    interval: number | undefined;
    circleRef: HTMLElement | undefined;

    constructor(props: Props) {
        super(props);
        this.state = { countdown: props.initialValue };
    }

    componentDidMount() {
        const { initialValue } = this.props;
        const id: any = setInterval(() => {
            this.update();
        }, 1000);
        this.interval = id as number;

        if (this.circleRef) {
            this.circleRef.style.animationDuration = initialValue + "s";
        }
    }

    update() {
        const { countdown } = this.state;
        const { onCountdownComplete } = this.props;

        if (countdown <= 1) {
            // countdown done
            this.interval && clearInterval(this.interval);
            onCountdownComplete();
        }

        // update
        this.setState({ countdown: countdown - 1 });
    }

    render(props: Props, state: State) {
        const { countdown } = state;
        const { radius } = props;

        return (
            <div class={`countdown ${styles["countdown"]}`}>
                <div class={`countdown-number ${styles["countdown-number"]}`}>{countdown}</div>
                <svg class={styles["svg"]}>
                    <circle
                        class={styles["circle"]}
                        ref={circle => (this.circleRef = circle as HTMLElement)}
                        r={radius! - 2}
                        cx={radius}
                        cy={radius}
                    />
                </svg>
            </div>
        );
    }
}
