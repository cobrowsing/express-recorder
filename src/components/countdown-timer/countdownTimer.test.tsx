import { render } from "preact-render-to-string";
import { CountdownTimer } from "./CountdownTimer";
import { h } from "preact";

it("Renders without crashing", () => {
    const div = document.createElement("div");
    render(
        <CountdownTimer
            initialValue={3}
            onCountdownComplete={() => {
                return;
            }}
        />,
        div
    );
});

it("Match snapshot", () => {
    const tree = render(
        <CountdownTimer
            initialValue={3}
            onCountdownComplete={() => {
                return;
            }}
        />
    );
    expect(tree).toMatchSnapshot();
});

console.log("I (test) render");
