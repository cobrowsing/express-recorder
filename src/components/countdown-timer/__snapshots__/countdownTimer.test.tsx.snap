// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown timer snapshot should render header with content 1`] = `"<div class=\\"countdown undefined\\"><div class=\\"countdown-number undefined\\">0</div><svg><circle r=\\"42\\" cx=\\"44\\" cy=\\"44\\"></circle></svg></div>"`;
